httpbakery/form/form_test.go

```diff
@@ -0,0 +1,219 @@
+}, {
+	about:       "error submitting form",
+	postError:   true,
+	expectError: `cannot get discharge from ".*": cannot start interactive session: cannot submit form: POST .*: httprequest: test error`,
+}, {
+	about:       "no schema",
+	emptySchema: true,
+	expectError: `cannot get discharge from ".*": cannot start interactive session: invalid schema`,
+}, {
+	about: "handler error",
+	handler: func(environschema.Fields) (map[string]interface{}, error) {
+		return nil, testError
+	},
+	expectError: `cannot get discharge from ".*": cannot start interactive session: cannot handle form: test error`,
+}}
+
+func (s *formSuite) TestFormLogin(c *gc.C) {
+	d := &formDischarger{}
+	d.discharger = bakerytest.NewLoginDischarger(nil, http.HandlerFunc(d.login))
+	defer d.discharger.Close()
+	d.discharger.Mux.Handle("/form", http.HandlerFunc(d.form))
+	svc, err := bakery.NewService(bakery.NewServiceParams{
+		Locator: d.discharger,
+	})
+	c.Assert(err, gc.IsNil)
+	for i, test := range formLoginTests {
+		c.Logf("%d. %s", i, test.about)
+		d.ignoreAccept = test.ignoreAccept
+		d.loginError = test.loginError
+		d.formUnsupported = test.formUnsupported
+		d.getError = test.getError
+		d.postError = test.postError
+		d.emptySchema = test.emptySchema
+
+		m, err := svc.NewMacaroon("", nil, []checkers.Caveat{{
+			Location:  d.discharger.Location(),
+			Condition: "test condition",
+		}})
+		c.Assert(err, gc.Equals, nil)
+		client := httpbakery.NewClient()
+		h := defaultHandler
+		if test.handler != nil {
+			h = test.handler
+		}
+		form.SetUpAuth(client, h)
+
+		ms, err := client.DischargeAll(m)
+		if test.expectError != "" {
+			c.Assert(err, gc.ErrorMatches, test.expectError)
+			continue
+		}
+		c.Assert(err, gc.IsNil)
+		c.Assert(len(ms), gc.Equals, 2)
+	}
+}
+
+func (s *formSuite) TestFormLoginNewRequestError(c *gc.C) {
+	client := httpbakery.NewClient()
+	form.SetUpAuth(client, defaultHandler)
+	u := url.URL{
+		Scheme: ":",
+	}
+	err := client.VisitWebPage(&u)
+	c.Assert(err, gc.ErrorMatches, "cannot create request: parse :://: missing protocol scheme")
+}
+
+type formDischarger struct {
+	discharger *bakerytest.LoginDischarger
+
+	ignoreAccept    bool
+	loginError      bool
+	formUnsupported bool
+	getError        bool
+	postError       bool
+	emptySchema     bool
+}
+
+func (d *formDischarger) login(w http.ResponseWriter, r *http.Request) {
+	if d.ignoreAccept {
+		w.Write([]byte("OK"))
+		return
+	}
+	if r.Header.Get("Accept") != "application/json" {
+		d.errorf(w, r, "bad accept header %q", r.Header.Get("Accept"))
+	}
+	if d.loginError {
+		httprequest.WriteJSON(w, http.StatusInternalServerError, testError)
+		d.discharger.FinishLogin(w, r, testError)
+		return
+	}
+	methods := map[string]string{}
+	if !d.formUnsupported {
+		r.ParseForm()
+		methods["form"] = d.discharger.Location() + "/form?waitid=" + r.Form.Get("waitid")
```

Comment:
Hmm, this seems like unwarranted chumminess with the LoginDischarger (it doesn't document waitid anywhere).

I wonder if LoginDischarger could provide a Context method
e.g.

    // Context returns the discharger request context from the
    // given request. It returns an empty string if no context is found.
    Context(req *http.Request) string

Then FinishLogin could be:

    // FinishLogin completes a login request. The context should
    // have been obtained from the original call to the login handler.
    // It panics if context is not valid.
    FinishLogin(w http.ResponseWriter, context string, err error)

Then we wouldn't need to know anything about waitid here.